tinygrad/runtime/ops_hip.py

```diff
@@ -26,13 +35,25 @@ def _copyin(self, x:np.ndarray): hip.hipMemcpyAsync_htod(self._buf, x.ctypes.dat
   def _copyout(self, x:np.ndarray): hip.hipMemcpy_dtoh(x.ctypes.data, self._buf, self.size * self.dtype.itemsize)
 
 class HIPProgram:
-  def __init__(self, name:str, prg:str, binary=False):
+  def __init__(self, name:str, prg:Union[str, bytes], binary=False):
     try:
       if not binary:
-        prog = hip.hiprtcCreateProgram(prg, name, [], [])
-        device_properties = hip.hipGetDeviceProperties(hip.hipGetDevice())
-        hip.hiprtcCompileProgram(prog, [f'--offload-arch={device_properties.gcnArchName}'])
-        prg = hip.hiprtcGetCode(prog)
+        assert isinstance(prg, str)
+        prg_hash = hashlib.sha256(prg.encode()).digest().hex()
+        prg_cache_path = os.path.join(hip_cache_dir(), prg_hash)
+        if os.path.exists(prg_cache_path):
+          with open(prg_cache_path, "rb") as f:
+            prg = f.read()
+        else:
+          prog = hip.hiprtcCreateProgram(prg, name, [], [])
+          device_properties = hip.hipGetDeviceProperties(hip.hipGetDevice())
+          hip.hiprtcCompileProgram(prog, [f'--offload-arch={device_properties.gcnArchName}'])
+          prg = hip.hiprtcGetCode(prog)
+          os.makedirs(hip_cache_dir(), exist_ok=True)
+          prg_cache_path_tmp = prg_cache_path + ".tmp"
```

Comment:
race condition, see the same issue in clang. in fact, this should share code with clang.

Reply:
> race condition, see the same issue in clang

Ok the multiprocess race condition should be resolved now.

> in fact, this should share code with clang.

Yeah generic binary caching should be pretty easy. The only annoying part is the compiler change detection. I could try to refactor this to be more generic and reusable if you'd like.